### ISSUE TITLE
Are there any plans to make fixed math completely deterministic?

### ISSUE THREAD

**Issue by zigzag312** [sungiant/abacus#21]
Hi,

Interesting project you got here. 

Are there any plans to make fixed point math & conversions completely deterministic? 

I found that `Fixed32.ToDouble` method uses `double` multiplication (which is floating point operation and is not deterministic across-platforms). Same for constructor that takes in a `double` value.

I need deterministic numerical type with math operations and conversions from/to .NET's floating point data types. Your project seems promising in that regard.

**Comment by zigzag312** [comment]
I have done some more research. Multiplying with a power of two FP numbers is deterministic after all. So there is no problem with `Fixed32.ToDouble` method.

But I found one `System.Math.Round` call inside constructor. I am not sure that method is deterministic (probably not I think).

**Comment by sungiant** [comment]
Hi Zigzag,

Sorry for the delay, I've not looked at this for ages.  It's not something I had considered by perhaps it would be worth doing if it makes sense.  Feel free to make a PR if you have changes.

My question for you is would it ever be possible to make the fixed point to floating point conversions deterministic?  I'm not convinced that we have any guarantee that the resultant floating point number would be stored with the exact same precision on all systems/architectures.  That concern was one reason for implementing fixed precision numbers in this project in the first place. ;-)

So, not sure about the above, but making the conversions deterministic on a the same system/architecture certainly makes sense.

**Comment by zigzag312** [comment]
Hi @sungiant,
  
Yes there is. If floating point number is multiplied/divided with power of two number, mantissa of FP number doesn't change. Only the exponent changes (point moves). 

> Floating point is a base-2 number. A number like 2^32 is represented as (exponent: 32, mantissa: 1). If you multiply this with another float (exp, man), the result is (exp + 32, man * 1). For division, the result is (expo - 32, man * 1). This is an exact result, so there's no rounding. If there's no rounding, rounding mode doesn't matter. Or to put it differently, multiplying the mantissa by 1 doesn't change the mantissa, so it doesn't matter how many bits it has.

I had lengthy discussion here: https://github.com/asik/FixedMath.Net/issues/8

I think that `Fixed32.ToDouble` method is deterministic, but `Fixed32` constructor is not due to the `System.Math.Round` call.

**Comment by sungiant** [comment]
Makes sense.  Though to what end if determinism is important?  You'd have to be super careful with that resultant floating point value and avoid running any floating point operations that are subject to variance across systems/architectures...

Still it sounds like the change is worth making.  I think I'll give this lib a bit of an overhaul at some point soon, it would be nice to finish the tests for fixed point numbers. ;-)
